### ISSUE TITLE
transformers, torch fail with triton installed, giving `OSError: could not get source code` in `triton` when importing `peft`

### ISSUE THREAD

**Issue by doradsoft** [Nuitka/Nuitka#2552]
When importing `peft` executable is created, but when running it, It throws:
`OSError: could not get source code`
and exits.

Full log:

```
False
/tmp/onefile_10990_1700048433_451408/bitsandbytes/cextension.py:34: UserWarning: The installed version of bitsandbytes was compiled without GPU support. 8-bit optimizers, 8-bit m
ultiplication, and GPU quantization are unavailable.
'NoneType' object has no attribute 'cadam32bit_grad_fp32'
Traceback (most recent call last):
  File "/tmp/onefile_10990_1700048433_451408/triton_reproduce.py", line 7, in <module>
    from peft import AutoPeftModelForCausalLM, AutoPeftModelForSeq2SeqLM

  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "/tmp/onefile_10990_1700048433_451408/peft/__init__.py", line 22, in <module peft>
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "/tmp/onefile_10990_1700048433_451408/peft/auto.py", line 30, in <module peft.auto>
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "/tmp/onefile_10990_1700048433_451408/peft/mapping.py", line 20, in <module peft.mapping>
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "/tmp/onefile_10990_1700048433_451408/peft/peft_model.py", line 39, in <module peft.peft_model>
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "/tmp/onefile_10990_1700048433_451408/peft/tuners/__init__.py", line 21, in <module peft.tuners>
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "/tmp/onefile_10990_1700048433_451408/peft/tuners/lora.py", line 42, in <module peft.tuners.lora>
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "/tmp/onefile_10990_1700048433_451408/bitsandbytes/__init__.py", line 16, in <module bitsandbytes>
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "/tmp/onefile_10990_1700048433_451408/bitsandbytes/nn/__init__.py", line 6, in <module bitsandbytes.nn>
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "/tmp/onefile_10990_1700048433_451408/bitsandbytes/nn/triton_based_modules.py", line 8, in <module bitsandbytes.nn.triton_based_modules>
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "/tmp/onefile_10990_1700048433_451408/bitsandbytes/triton/dequantize_rowwise.py", line 10, in <module bitsandbytes.triton.dequantize_rowwise>
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "/tmp/onefile_10990_1700048433_451408/triton/__init__.py", line 29, in <module triton>
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "/tmp/onefile_10990_1700048433_451408/triton/language/__init__.py", line 8, in <module triton.language>
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "/tmp/onefile_10990_1700048433_451408/triton/language/libdevice.py", line 4, in <module triton.language.libdevice>
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "/tmp/onefile_10990_1700048433_451408/triton/language/core.py", line 1175, in <module triton.language.core>
  File "/tmp/onefile_10990_1700048433_451408/triton/runtime/jit.py", line 419, in jit
  File "/tmp/onefile_10990_1700048433_451408/triton/runtime/jit.py", line 412, in decorator
  File "/tmp/onefile_10990_1700048433_451408/triton/runtime/jit.py", line 306, in __init__
  File "/tmp/onefile_10990_1700048433_451408/inspect.py", line 1262, in getsource
  File "/tmp/onefile_10990_1700048433_451408/inspect.py", line 1244, in getsourcelines
  File "/tmp/onefile_10990_1700048433_451408/inspect.py", line 1081, in findsource
OSError: could not get source code
```

I believe that the relevant source line is:
https://github.com/openai/triton/blob/728282b11eebca01a64853d251b9fa9af0afe8f1/python/triton/runtime/jit.py#L585C44-L585C53

Reproduce details as guided:
* Nuitka version, full Python version, flavor, OS, etc. as output by this exact command.

```txt
1.9rc6
Commercial: 2.3.2
Python: 3.11.6 (main, Oct 19 2023, 12:55:47) [GCC 5.4.0 20160609]
Flavor: pyenv
Executable: /root/.pyenv/versions/3.11.6/bin/python
OS: Linux
Arch: x86_64
Distribution: Ubuntu (based on Debian) 16.04.7
Version C compiler: /usr/lib/ccache/gcc (gcc 5.4.0).
```
We must use staging version until some other issues are merged.

* How did you install Nuitka and Python
`pip install Nuitka-commercial-staging.zip`
no virtualenv.

<details>
  <summary>The specific PyPI names and versions</summary>
  ```
accelerate==0.20.3
aiohttp==3.8.5
aiosignal==1.3.1
async-timeout==4.0.3
attrs==23.1.0
backoff==2.2.1
bitsandbytes==0.41.1
certifi==2023.7.22
charset-normalizer==3.2.0
click==8.1.7
cmake==3.27.7
datasets==2.14.5
Deprecated==1.2.14
dill==0.3.7
einops==0.6.1
filelock==3.12.4
frozenlist==1.4.0
fsspec==2023.6.0
googleapis-common-protos==1.60.0
grpc-interceptor==0.15.3
grpcio==1.58.0
grpcio-reflection==1.58.0
grpcio-status==1.58.0
grpcio-tools==1.51.1
hf_transfer==0.1.3
huggingface-hub==0.16.4
idna==3.4
Jinja2==3.1.2
lit==17.0.5
loguru==0.6.0
MarkupSafe==2.1.3
mpmath==1.3.0
multidict==6.0.4
multiprocess==0.70.15
mypy-protobuf==3.4.0
networkx==3.1
Nuitka @ file:///python-packages/Nuitka-commercial-staging.zip#sha256=77333731fa0274fbabcc4f7a959c57903a7bd800e563f761ed318f0c54676c5e
numpy==1.26.0
nvidia-cublas-cu11==11.10.3.66
nvidia-cuda-cupti-cu11==11.7.101
nvidia-cuda-nvrtc-cu11==11.7.99
nvidia-cuda-runtime-cu11==11.7.99
nvidia-cudnn-cu11==8.5.0.96
nvidia-cufft-cu11==10.9.0.58
nvidia-curand-cu11==10.2.10.91
nvidia-cusolver-cu11==11.4.0.1
nvidia-cusparse-cu11==11.7.4.91
nvidia-nccl-cu11==2.14.3
nvidia-nvtx-cu11==11.7.91
opentelemetry-api==1.15.0
opentelemetry-exporter-otlp==1.15.0
opentelemetry-exporter-otlp-proto-grpc==1.15.0
opentelemetry-exporter-otlp-proto-http==1.15.0
opentelemetry-instrumentation==0.36b0
opentelemetry-instrumentation-grpc==0.36b0
opentelemetry-proto==1.15.0
opentelemetry-sdk==1.15.0
opentelemetry-semantic-conventions==0.36b0
ordered-set==4.1.0
packaging==23.1
pandas==2.1.1
peft==0.4.0
Pillow==10.0.1
protobuf==4.24.3
psutil==5.9.5
pyarrow==13.0.0
python-dateutil==2.8.2
pytz==2023.3.post1
PyYAML==6.0.1
regex==2023.8.8
requests==2.31.0
safetensors==0.3.3
scipy==1.11.2
sentencepiece==0.1.99
six==1.16.0
sympy==1.12
-e git+ssh://git@github.com/huggingface/text-generation-inference.git@a5def7c222174e03d815f890093584f3e815c5ce#egg=text_generation_server&subdirectory=server
texttable==1.6.7
tokenizers==0.13.3
torch==2.0.1
tqdm==4.66.1
transformers==4.33.2
triton==2.0.0
typer==0.6.1
types-protobuf==4.24.0.4
typing_extensions==4.8.0
tzdata==2023.3
urllib3==2.0.5
wrapt==1.15.0
xxhash==3.3.0
yarl==1.9.2
zstandard==0.22.0
  ```python
</details>

Other programs do work. even those which includes `datasets` and `evaluate`.
* Also supply a Short, Self Contained, Correct, Example
```py
import os
import json
from loguru import logger
import torch

from transformers import AutoTokenizer
from peft import AutoPeftModelForCausalLM, AutoPeftModelForSeq2SeqLM


def download_and_unload_peft(model_id):
    torch_dtype = torch.float16

    logger.info("Peft model detected.")
    logger.info("Loading the model it might take a while without feedback")
    try:
        model = AutoPeftModelForCausalLM.from_pretrained(
            model_id,
            torch_dtype=torch_dtype,
            trust_remote_code=True,
            low_cpu_mem_usage=True,
        )
    except Exception:
        model = AutoPeftModelForSeq2SeqLM.from_pretrained(
            model_id,
            torch_dtype=torch_dtype,
            trust_remote_code=True,
            low_cpu_mem_usage=True,
        )
    logger.info("Loaded.")
    logger.info("Merging the lora weights.")

    base_model_id = model.peft_config["default"].base_model_name_or_path

    model = model.merge_and_unload()

    os.makedirs(model_id, exist_ok=True)
    cache_dir = model_id
    logger.info(f"Saving the newly created merged model to {cache_dir}")
    tokenizer = AutoTokenizer.from_pretrained(base_model_id)
    model.save_pretrained(cache_dir, safe_serialization=True)
    model.config.save_pretrained(cache_dir)
    tokenizer.save_pretrained(cache_dir)


download_and_unload_peft("ybelkada/opt-350m-lora")
```
* Provide in your issue the Nuitka options used

```sh
python -m nuitka --main=triton_reproduce.py --standalone --onefile --static-libpython=no --include-distribution-met
adata=accelerate --include-distribution-metadata=bitsandbytes --include-distribution-metadata=datasets --include-distribution-metadata=jinja2 --include-distribution-metadata=pand
as --include-distribution-metadata=psutil --include-distribution-metadata=safetensors --include-distribution-metadata=torch --include-distribution-metadata=tqdm --include-distrib
ution-metadata=transformers --include-distribution-metadata=huggingface-hub --include-distribution-metadata=scipy --include-distribution-metadata=tokenizers --include-distributio
n-metadata=protobuf --include-distribution-metadata=peft --noinclude-unittest-mode=allow

```

**Comment by kayhayen** [comment]
After I installed `triton` into my living installation, where I do all my debugging, it became apparent that it's woven into many parts of pytorch, but entirely optionally. So I reproduced this with the other example suddenly all the time, and finally managed to get it completely eradicated with anti-bloat measures. I doubt your example is actually intending to use triton, so it should work with factory and staging.

I am having issues with `evaluate` which I will report in the other issue later, that I want to solve first.

**Comment by kayhayen** [comment]
Thanks for your report, this is fixed on the factory branch, which is a development version under rapid development. You can try it out by going here: https://nuitka.net/doc/factory.html 

Feedback if this is working is very welcome, just please do not share plans of doing it, but rather confirmations or denials of it working.

**Comment by doradsoft** [comment]
I ran it again on latest staging and it fails:

> Nuitka-Options:INFO: Used command line options: --main=triton\_reproduce.py --standalone --onefile --static-libpython=no --include-distribution-metadata=accelerate -
> -include-distribution-metadata=bitsandbytes --include-distribution-metadata=datasets --include-distribution-metadata=jinja2 --include-distribution-metadata=pandas -
> -include-distribution-metadata=psutil --include-distribution-metadata=safetensors --include-distribution-metadata=torch --include-distribution-metadata=tqdm --inclu
> de-distribution-metadata=transformers --include-distribution-metadata=huggingface-hub --include-distribution-metadata=scipy --include-distribution-metadata=tokenize
> rs --include-distribution-metadata=protobuf --include-distribution-metadata=peft --noinclude-unittest-mode=allow
> Nuitka:INFO: Starting Python compilation with Nuitka '1.9rc9' on Python '3.11' commercial grade '2.3.2'.
> PASS 1: 96.2%|████████████████████████ | 2449/2547, transformers.integrationsTraceback (most recent call last):
>   File "/root/.pyenv/versions/3.11.6/lib/python3.11/site-packages/nuitka/plugins/Plugins.py", line 82, in withPluginProblemReporting
>     yield
>   File "/root/.pyenv/versions/3.11.6/lib/python3.11/site-packages/nuitka/plugins/Plugins.py", line 534, in considerImplicitImports
>     cls.\_considerImplicitImports(plugin=plugin, module=module)
>   File "/root/.pyenv/versions/3.11.6/lib/python3.11/site-packages/nuitka/plugins/Plugins.py", line 426, in \_considerImplicitImports
>     for full\_name in iterateModuleNames(plugin.getImplicitImports(module)):
>   File "/root/.pyenv/versions/3.11.6/lib/python3.11/site-packages/nuitka/plugins/Plugins.py", line 403, in iterateModuleNames
>     for v in value:
>   File "/root/.pyenv/versions/3.11.6/lib/python3.11/site-packages/nuitka/plugins/standard/TransformersPlugin.py", line 247, in getImplicitImports
>     for sub\_module\_name in self.queryRuntimeInformationSingle(
>                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/root/.pyenv/versions/3.11.6/lib/python3.11/site-packages/nuitka/plugins/PluginBase.py", line 1128, in queryRuntimeInformationSingle
>     return self.queryRuntimeInformationMultiple(
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/root/.pyenv/versions/3.11.6/lib/python3.11/site-packages/nuitka/plugins/PluginBase.py", line 1090, in queryRuntimeInformationMultiple
>     feedback = check\_output([sys.executable, "-c", cmd])
>                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/root/.pyenv/versions/3.11.6/lib/python3.11/site-packages/nuitka/utils/Execution.py", line 141, in check\_output
>     raise NuitkaCalledProcessError(exit\_code, cmd, output=output, stderr=stderr)
> nuitka.utils.Execution.NuitkaCalledProcessError: Command '['/root/.pyenv/versions/3.11.6/bin/python', '-c', '\\\nfrom \_\_future\_\_ import print\_function\nfrom \_\_futur
> e\_\_ import absolute\_import\n\ntry:\n   import transformers.integrations\nexcept ImportError:\n    import sys\n    sys.exit(38)\nprint(repr(list(transformers.integra
> tions.\_import\_structure.keys())))\nprint("-" * 27)\n']' returned non-zero exit status 1. Error was b'Traceback (most recent call last):\n  File "<string>", line 10,
>  in <module>\nAttributeError: module \'transformers.integrations\' has no attribute \'\_import\_structure\''.
> FATAL: transformers: Plugin issue while working on 'module 'transformers.integrations''. Please report the bug with the above traceback included.
> Nuitka-Reports:INFO: Compilation report written to file 'nuitka-crash-report.xml'.
> Nuitka-Reports:INFO: Compilation crash report written to file 'nuitka-crash-report.xml'.

[nuitka-crash-report.zip](https://github.com/Nuitka/Nuitka/files/13413517/nuitka-crash-report.zip)

**Comment by doradsoft** [comment]
Latest crash I think was due to some environment-specific issue. Sorry if you already invested on this one. I'm keeping it for now as compilation is still running so I'm not 100% it was solved. As soon as I saw that it's proceeding more than before I just wanted to stop you from wasting your time. Will update nonetheless.

**Comment by kayhayen** [comment]
This is a weakness of this plugin, that it lists these modules, and I am not aware they would have to be all actually do it for specific transformers releases. It appears you are using 4.33.2 and I was using 4.35.1, but do not jump to changing it, rather I will make sure it works for both. The transformers plugin obviously has to allow for this lazy loading to be done or not or else we end up supporting only very specific releases of it.

**Comment by kayhayen** [comment]
So, I have updated both factory and staging branches with a fix for this, no longer requires lazy loaders in the hard coded list, at some point, the lazy loading mechanism should move towards what we have done for `huggingface_hub` and `lazy` but I am not there yet.

**Comment by snps-dorad** [comment]
1. I confirm that compilation works again. I'll update regarding runtime.
2. 
> no longer requires lazy loaders in the hard coded list
I'm not sure what that means. I assumed you referred the `--include-distribution-metadata` args so I dropped them and it fails on runtime with `Nuitka: Need to use this as an option to compile with '--include-distribution-metadata=bitsandbytes'`

**Comment by doradsoft** [comment]
1. I confirm that compilation works again.
2. 
> no longer requires lazy loaders in the hard coded list

I'm not sure what that means. I assumed you referred the `--include-distribution-metadata` args so I dropped them and it fails on runtime with `Nuitka: Need to use this as an option to compile with '--include-distribution-metadata=bitsandbytes'`
3. After re compiling with all '--include-distribution-metadata`s, original runtime error persists 😢:

**Comment by kayhayen** [comment]
Yes, I miswrote that. I think the metadata checks kind of also depend on what's installed, not sure how to get at a full list of what it attempts to use without looking at not-found imports, all of them, which is unreasonably laborsome, so I will just add them as we encounter them.

I do however recall, that this may actually be a case of a function that if we were to inline it, would make it visible at compile time, so I am somewhat optimistic, we might end up seeing these calls eventually in an automated way once this gets added.

**Comment by kayhayen** [comment]
The good news, this makes it into 1.9 proper, and I kind of expect it to happen today. I will get in touch with you to assess open things from yours side @doradsoft later this week.

**Comment by kayhayen** [comment]
So, this is part of the 1.9 release, please open a new issue, if you find anything not perfect with `transformers` import still.

**Comment by snps-dorad** [comment]
> So, this is part of the 1.9 release, please open a new issue, if you find anything not perfect with `transformers` import still.

I'm not sure I get it, the original issue persists (on stable 1.9 commercial as well), should I just copy the entire issue contents but with different `Nuitka --version` output?

**Comment by doradsoft** [comment]
> So, this is part of the 1.9 release, please open a new issue, if you find anything not perfect with `transformers` import still.

I'm not sure I get it, the original issue persists (on stable 1.9 commercial as well), should I just copy the entire issue contents but with different `Nuitka --version` output?

**Comment by doradsoft** [comment]
> So, this is part of the 1.9 release, please open a new issue, if you find anything not perfect with `transformers` import still.

I'm not sure I get it, the original issue persists (on stable 1.9 commercial as well), should I just copy the entire issue contents but with different `Nuitka --version` output?

**Comment by kayhayen** [comment]
I am currently running the reproducer in a loop, and doing more anti-bloat work for triton in torch, and IPython in loguru, more probably needed.

**Comment by kayhayen** [comment]
So, transformers is asking me for more metadata to be included now, this test case clearly covers more than others in terms of what's used, but we are getting there I guess. I am trying to have 1.9.2 released with this, assuming it is only about this kind of work.

**Comment by kayhayen** [comment]
So, 1.9.2 is out, and the example you gave compiled for me. I am still going to continue the loop, I guess with `--include-package=torch` for coverage, as probably there are yet more modules using `triton` still, but that can be 1.9.3 material.

For transformers a large amount of meta data packages was added, let me know if you find one missing, you should be able to use none at all now.

**Comment by snps-dorad** [comment]
> So, 1.9.2 is out, and the example you gave compiled for me
The original problem was about runtime error, not compilation

> you should be able to use none at all now.
Excellent, it wasn't complain about any missing distribution!

However, the original runtime error persists with the exact same traceback😢

**Comment by kayhayen** [comment]
@doradsoft with the provided example, really? Or is this mixing with the `evaluate` issue, which I have yet to do, and am aiming for in 1.9.3 now. I might have to do --include-package=torch to be sure, I am not missing any uses there, and then we probably need to be very specific about the versions. The output of --report would tell me which exactly you are using. I will however first try and become more complete in torch with those. Now that I see it, `peft` imports `triton` via `bitsandbytes` bit I didn't see that one at all, so I will your reproduce with an `--include-package=bitsandbytes` as well I guess. Maybe also a version thing.

**Comment by kayhayen** [comment]
> FATAL: Error, failed to locate package 'bitsandbytes' you asked to include.

Ah yes, optional dependencies, got to love these people really hard.

**Comment by kayhayen** [comment]
So, I am compiling with `bitsandbytes` and finding and removing `triton` usages there as well, not sure how many iterations it will take, but hopefully it's not too bad.

**Comment by doradsoft** [comment]
> @doradsoft with the provided example, really? Or is this mixing with the `evaluate` issue

Yes. The exact same code and environment, not mixing with #2529. 

Thanks for handling this.

**Comment by kayhayen** [comment]
I have successfully removed `triton` from `bitsandbytes` as well, and found missing DLLs for it, that I am still working on, somehow maybe there is a bug, with my DLL pattern not matching although they should.

**Comment by kayhayen** [comment]
The example works, with staging and factory, I will soon release 1.9.3, but there is another issue that I want to get it, but should be tomorrow.

**Comment by doradsoft** [comment]
Thanks for treating this! 
1. I confirm that the specific error is gone! 🕺so you can close the issue now.
2. There is another runtime error with the same reproduce.
3. As this reproduce was made as a MCVE of a real code, and that specific error is gone for the real code as well I don't think I'll open another issue for the new runtime error, as for now it doesn't reproduce for the real code.
4. However, there is a new runtime error only for the real code. And for that I'll now work on a new MCVE and open a new issue.

**Comment by kayhayen** [comment]
@doradsoft you are welcome to through all the MVCE at me that you can come up with. Closing this one. Symptoms and error causes do not relate a lot. I forgot to add the factory tag, this one is solved in 1.9.3 supposedly and I released it the other day.
